File: 0048-rotate-image/0048-rotate-image.java
class Solution {
    public void rotate(int[][] matrix) {
        int n = matrix[0].length-1;
//         transpose matrix
        for(int i=0; i<matrix.length; i++){
            for(int j=0; j<matrix[0].length; j++){
                if(i>j){
                    int temp = matrix[i][j];
                    matrix[i][j]=matrix[j][i];
                    matrix[j][i]=temp;
                }
            }
        }
        
//         swap cols from middle
        for(int i=0;i<matrix.length; i++){
            for(int j=0; j<matrix[0].length/2; j++){
                    int temp = matrix[i][j];
                    matrix[i][j]=matrix[i][n-j];
                    matrix[i][n-j]=temp;            }
        }
    }
}

File: 0048-rotate-image/0048-rotate-image.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        int n = matrix.size();

        for(int i = 0; i < n / 2; i++) { // layers
            for(int j = i; j < n - i - 1; j++) {

                int temp = matrix[i][j]; // top

                // left -> top
                matrix[i][j] = matrix[n - 1 - j][i];

                // bottom -> left
                matrix[n - 1 - j][i] = matrix[n - 1 - i][n - 1 - j];

                // right -> bottom
                matrix[n - 1 - i][n - 1 - j] = matrix[j][n - 1 - i];

                // top -> right
                matrix[j][n - 1 - i] = temp;
            }
        }
    }
};